### DISCUSSION TITLE
Extension Development Experience

### DISCUSSION THREAD

**Discussion by FredericLatour** [directus/directus#18050] [Archive]
Hi,
While evaluating Directus, I decided to develop a custom endoint.

I scaffolded an endpoint extension (typescript).
I roughly got some result using the recipe example :

![image](https://user-images.githubusercontent.com/1407847/229859926-23846cfc-d172-448a-9f55-50fde7eb3fa5.png)

Isn't there however the possibility to have typescript types for the following ex?
- services — All API internal services.
- exceptions — API exception objects that can be used to throw "proper" errors.
- emitter — [Event emitter](https://github.com/directus/directus/blob/main/api/src/emitter.ts) instance that can be used to trigger custom events for other extensions

What about `req` and `res`  types ? Why not providing types ?

When using Knex directly through `database` (from the extension context object), is there a way to get typescript types for the tables ?

Did I miss something ?
Any info appreciated. Thanks.

**Comment by FredericLatour** [comment]
No one is able to let me know if I'm missing something or if it is indeed the current situation on the typescript side of things when developing custom endpoint extensions?

**Comment by mahgoh** [comment]
I had a similar experience when first starting - it cost me quite some time to understand and find all the different types by looking through the source code. It has been a few months since I wrote my last extension, so this might have improved. As far as I know, there's no out-of-the-box support for this.

The documentation only briefly explains parts of it and doesn't go too much into detail - which forced me to look at the source code. 

Concerning `req` and `res`: The router is basically an `express` router with additional properties. Installing `@types/express` partially solves this issue by providing the express types. However, additional properties such as `schema` or `accountability` are missing.

```bash
# install express types as dev dependency
npm i -D @types/express
```

Complete type support would help **tremendously** - especially to get started with extension development. In my opinion, this is a make-or-break feature. Instead of troubleshooting types, developers could focus on the actual extension and build more awesome stuff.

@rijkvanzanten is this a technical difficulty or just something not done yet? I think this would enormously benefit the developer experience.

**Reply by rijkvanzanten** [reply]
Bit of both! The extensions are installed in `dist` of Directus, and we're not shipping the source code (eg `.ts` files) in the bundles of the `directus` package itself, which in turn means that those types aren't really available anywhere. That being said, we did just start `@directus/types` to solve that 🙂

**Reply by mahgoh** [reply]
I see, great to hear that 🔥 Thanks for your reply and the great work

**Comment by FredericLatour** [comment]
ok, it's not just me then :)
Considering the chore product is written in TS, It's a bit weird that most types are not provided.

**Reply by rijkvanzanten** [reply]
> Considering the chore product is written in TS, It's a bit weird that most types are not provided.

It's all around source code vs dist. The `dist` bundle of the `directus` package doesn't include the source code. `@directus/types` is in the works to help here 🙂

**Comment by echocrow** [comment]
ran into this too in a first hook extension test, coincidentally also while evaluating Directus as a potential project fit. personally I found accurate types incredibly powerful, particularly when first discovering an API/SDK.

sounds like previous architecture changes made typing these extension functions difficult, but it looks like that could now (or soon) be overcome via the aforementioned `@directus/types` helper package(?)

some (hopefully not too presumptuous) thoughts on where this could go:
- with reusable types in place, a low-hanging fruit would be to get more types in place for the various `context` objects
  - particularly the shared `services` and `exceptions` props, currently typed as `any`, may be yield good type-bang for the typing-buck
- the `ItemsService` is probably a heavily used utility in extensions. if we can define item types at a higher level, the collection name could potentially be typed, and the shape of the returned items be inferred.
  - this could perhaps be achieved similarly to how the [standard SDK](https://docs.directus.io/reference/sdk.html#typescript) allows type injections via a `TypeMap` generic. _(other methods are available.)_
- looking at hook extensions specifically, there is a limited set of valid events for `filter()`, `action()`, etc.; building on the above, this could be taken as far as typing and auto-completing:
  - known, static events, such as `auth.login`, `server.start`, `cli.before`, `items.read`
  - schema-specific events, such as `<collection>.items.read`
  - inferring the (partial) shape of `payload` inside filter & action hooks, based on the `<collection>` name

bringing this all together, a fictitious example to illustrate what this _could_ look like—with type inference notes:
```ts
import { defineHook } from '@directus/extensions-sdk';

// (schema items map)
// (pretend this is auto-generated elsewhere and imported here)
interface MyProduct {
  id: number;
  type: string;
  hs_code: string;
  // (draw the rest of the owl)
}
interface MySchema {
  product: MyProduct;
}
// (end of schema items map)

export default defineHook<MySchema>(({ filter }, { services }) => {
  const { ItemsService } = services;
  // ^ services typed & auto-completed

  filter('product.items.update', async (payload, meta, context) => {
    // ^ event name auto-completed from `prod...` to
    // `product.items.read`, `product.items.update`, ...

    if (payload.hs_code) return;
    // ^ inferred as Partial<MyProduct>

    const { schema, accountability, database: knex } = context;
    // ^ context typed & auto-completed

    const itemsService = new ItemsService(
      meta.collection, // < inferred as literal string 'product'
      { schema, accountability, knex }
    );
    // ^ inferred as ItemsService<MyProduct>

    const item = await itemsService.readOne(meta.keys[0]);
    // ^ inferred as MyProduct

    if (!item.hs_code) payload.hs_code = myDefaultHsCode(item.type);
  });
});
```

thoughts? or is this pushing type inference too far?

**Reply by echocrow** [reply]
@rijkvanzanten if the Directus team is cool with this direction (and work is not already underway) I could look into working towards this sometime in the next few weeks. (unsure if I can contribute to a sufficient degree, as I'm new to the code base, but happy to give it a shot.)

**Reply by abdonrd** [reply]
Sounds really good!

**Reply by echocrow** [reply]
(addendum to the example I had shared above that may be a but more controversial)
- `defineHook` and other extension functions currently already use a TypeScript generic. IDK if the current generic adds any meaningful value (at least for hooks and endpoints it does not seem to?). In any case, either this would be a breaking change, or another way of passing in schema types may be necessary.
  - just to list one other method: some packages solve this via a type import to a stub that can be re-wired via tsconfig by the end-user. in this case, the type map would not need to be passed in as a generic anywhere. instead, the tsconfig would include a path such as `"@directus/schema": "path/to/my-generated-types.d.ts"`. This can be _convenient_ solution, but—being a "clever" trick—is also a lot less explicit.
- typing `payload` in filter & action item hooks may be _technically incorrect_. I didn't test is, but I suspect `payload` contains raw user input, meaning it's subject to type validation, and `unknown` is already the correct starting type.
  - I'd assume thought that payload` would most of the time effectively be a partial of the corresponding collection type, especially when the majority of events come from the studio or an (equally typed) SDK. so typing it as partial may be a lie, but a pretty good and convenient one.

**Reply by nickrum** [reply]
That looks exactly like what I had in mind as a long term goal for types in API extensions.

There have already been attempts at improving the situation, but this always led to a lot of duplication between the API and `@directus/types` (which was previously `@directus/shared`) because the services are tightly coupled with the API itself. That's why we didn't pursue this direction further. Ideally, we would be able to decouple thoses services from the API and put them into their on package. That way the Extensions SDK could import it and provide the types of the services that way. This is a much bigger long term task, though.

**Reply by echocrow** [reply]
> led to a lot of duplication between the API and @directus/types (which was previously @directus/shared) because the services are tightly coupled with the API itself.

yeah I noticed and feared the same when I poked around the current repo for this. I also figured _ideally_ these types would be co-located with the implementation (e.g. in their own package), avoiding type drifting and code duplication. as you pointed out though, that is a bigger refactor.

I assume you already considered, as an interim solution, to import types from `@directus/types` into the API package to type parts of the implementation in the API? It's still a band-aid compared to co-location, but potentially a longer-hanging fruit for now. That said, I may also missing a crucial root cause of code duplication.

**Reply by nickrum** [reply]
The last attempt was at #15711. While it was a noteworthy attempt, it unfortunately simply felt too verbose and too much of a maintenancy burden.
What do you think how such an interim solution should look like?

**Comment by luochuanyuewu** [comment]
Similarly in my circumstance, if the developer experience of Directus is elevated, I am confident that it will attract the participation of developers, leading to increased popularity and the gradual flourishing of its ecosystem.
